cmake_flags_and_output.md

```diff
@@ -0,0 +1,237 @@
+This can speed up the linking process, but produces a binary that cannot be debugged.
+In that case, prefer explicitly raising a warning telling the developer that he may be doing something wrong. 
+Also, such options should be disabled if applies.
+
+Bad:
+```cmake
+option(STRIP_DEBUG_SYMBOLS_FUNCTIONS
+    "Do not generate debugger info for ClickHouse functions.
+    ${STRIP_DSF_DEFAULT})
+
+if (STRIP_DEBUG_SYMBOLS_FUNCTIONS)
+    target_compile_options(clickhouse_functions PRIVATE "-g0")
+endif()
+
+```
+Better:
+
+```cmake
+# Provides faster linking and lower binary size.
+# Tradeoff is the inability to debug some source files with e.g. gdb
+# (empty stack frames and no local variables)."
+option(STRIP_DEBUG_SYMBOLS_FUNCTIONS
+    "Do not generate debugger info for ClickHouse functions."
+    ${STRIP_DSF_DEFAULT})
+
+if (STRIP_DEBUG_SYMBOLS_FUNCTIONS)
+    message(WARNING "Not generating debugger info for ClickHouse functions")
+    target_compile_options(clickhouse_functions PRIVATE "-g0")
+endif()
+```
+
+### In the option's description, explain WHAT the option does rather than WHY it does something.
+
+The WHY explanation should be placed in the comment.
+You may find that the option's name is self-descriptive.
+
+Bad:
+
+```cmake
+option(ENABLE_THINLTO "Enable Thin LTO. Only applicable for clang. It's also suppressed when building with tests or sanitizers." ON)
+```
+
+Better:
+
+```cmake
+# Only applicable for clang.
+# Turned off when building with tests or sanitizers.
+option(ENABLE_THINLTO ON).
+```
+
+### Don't assume other developers know as much as you do.
+
+In ClickHouse, there are many tools used that an ordinary developer may not know. If you are in doubt, give a link to
+the tool's docs. It won't take much of your time.
+
+Bad:
+
+```cmake
+option(ENABLE_THINLTO "Enable Thin LTO. Only applicable for clang. It's also suppressed when building with tests or sanitizers." ON)
+```
+
+Better (combined with the above hint):
+
+```cmake
+# https://clang.llvm.org/docs/ThinLTO.html
+# Only applicable for clang.
+# Turned off when building with tests or sanitizers.
+option(ENABLE_THINLTO ON).
+```
+
+Other example, bad:
+
+```cmake
+option (USE_INCLUDE_WHAT_YOU_USE "Use 'include-what-you-use' tool" OFF)
+```
+
+Better:
+
+```cmake
+# https://github.com/include-what-you-use/include-what-you-use
+option (USE_INCLUDE_WHAT_YOU_USE)
+```
+
+### Prefer consistent default values.
+
+CMake allows you to pass a plethora of values representing boolean `true/false`, e.g. `1, ON, YES, ...`.
+Prefer the `ON/OFF` values, if possible.
+
+
+## List of CMake flags
+
+* This list is auto-generated by [this bash script](cmake_flags_md_generator.sh).
+* The flag name is a link to its position in the code.
+
+| Name | Default value | Description | Comment |
+|------|---------------|-------------|---------|
+| [`ENABLE_IPO`](http://github.com/clickhouse/clickhouse/blob/master//CMakeLists.txt) | `` | Enable full link time optimization |  |
```

Comment:
Let's sort the list. Or at least group all ENABLE, USE options.

Reply:
Ok, so I'll slighly change the document structure.